Allow PVs to specify supplemental GIDs

wongma7 opened the pull request:
Retry of https://github.com/kubernetes/kubernetes/pull/28691 . Adds a Kubelet helper function for getting extra supplemental groups

childsb commented:
@k8s-bot ok to test

rootfs commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -227,12 +221,34 @@ func (vm *volumeManager) Run(stopCh <-chan struct{}) {
+		}
+	}
+
+	result := make([]int64, 0, supplementalGroups.Len())
```
what if len == 0

wongma7 replied:
It will return empty slice, I can make it return it before this line

pmorie replied:
That's the compiler's job

rootfs commented on pkg/securitycontext/provider.go:
```diff
@@ -47,27 +47,30 @@ func (p SimpleSecurityContextProvider) ModifyContainerConfig(pod *api.Pod, conta
+			hostConfig.GroupAdd = append(hostConfig.GroupAdd, strconv.Itoa(int(group)))
+		}
+
+		if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.FSGroup != nil {
```
can be consolidated into the first `pod.Spec.SecurityContext != nil `

wongma7 replied:
ok, thought the append order was deliberate

rootfs commented:
@wongma7 have you tried the previous failed slow e2e tests?

wongma7 commented:
@rootfs yes, they pass now

pmorie commented on pkg/kubelet/container/helpers.go:
```diff
@@ -44,6 +44,7 @@ type RuntimeHelper interface {
 	GetClusterDNS(pod *api.Pod) (dnsServers []string, dnsSearches []string, err error)
 	GetPodDir(podUID types.UID) string
 	GeneratePodHostNameAndDomain(pod *api.Pod) (hostname string, hostDomain string, err error)
+	GetExtraSupplementalGroupsForPod(pod *api.Pod) []int64
```
I don't think this needs to be in the runtime helper interface -- why not just a utility function?

wongma7 replied:
Wasn't sure where to put this, where would I make the utility function? (before, the volumemanager was passed to the runtime but I changed it because now volumemanager itself needs an initialized runtime)

pmorie replied:
Nevermind, I get why this is in there.  Be patient with me while I spin my brain back up.

pmorie replied:
Might as well put a doc comment on this declaration.

pmorie commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -510,20 +510,23 @@ func verifyNonRoot(app *appctypes.App, ctx *api.SecurityContext) error {
 }
 
-func setSupplementaryGIDs(app *appctypes.App, podCtx *api.PodSecurityContext) {
+func setSupplementaryGIDs(app *appctypes.App, podCtx *api.PodSecurityContext, supplementalGids []int64) {
```
call this `setSupplementalGIDs`

wongma7 replied:
'supplementary' is used throughout the rkt code, should I change all of them

pmorie commented on pkg/kubelet/rkt/rkt_test.go:
```diff
@@ -937,24 +937,26 @@ func TestSetApp(t *testing.T) {
 	nonRootUser := int64(42)
 	runAsNonRootTrue := true
-	fsgid := int64(3)
+	fsgid := int64(4)
```
why did this change?

wongma7 replied:
Wanted to match the append order i.e. when appending groups put fsgroup last, so I changed the test. I'll change the test back and change the implementation instead

pmorie replied:
You don't have to change it back, i just wondered.

pmorie commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -227,12 +221,38 @@ func (vm *volumeManager) Run(stopCh <-chan struct{}) {
+		}
+	}
+
+	if supplementalGroups.Len() == 0 {
```
there is not a lot of value in this optimization, remove it

lavalamp commented:
Please verify that https://github.com/kubernetes/kubernetes/issues/29224 isn't related to this PR before LGTMing.

wongma7 commented:
squashed and fixed a problem in the rkt code so it behaves just like docker https://github.com/kubernetes/kubernetes/pull/29119/commits/d914f5b6765524125c5af664dd6c8c274579bf11#diff-717f367da08d0bb8eb8c45d38f164eb8R513

& still verifying the e2e flake problem

childsb commented:
Just to recap, the e2e test now passes and Issue #29224 is verified not to be related to this PR.  Going to look for a final review from @pmorie

pmorie commented:
@yifan-gu for rkt changes

pmorie commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -395,32 +384,26 @@ func getExpectedVolumes(pod *api.Pod) []string {
-		}
+// getExtraSupplementalGid returns the value of an extra supplemental GID as
+// defined by an annotation on a volume and a boolean indicating whether the
+// volume defined a GID.
```
I think this is whether the volume defines a GID that the pod doesn't already request, accurate?

wongma7 replied:
yes, added that for clarity

pmorie commented on pkg/kubelet/volumemanager/volume_manager_test.go:
```diff
@@ -0,0 +1,276 @@
+
+		err = manager.WaitForAttachAndMount(pod)
+		if err != nil {
+			t.Errorf("Expected success: %v", err)
```
you should do a `continue` here -- otherwise you'll just get an extra failure

pmorie commented on pkg/kubelet/volumemanager/volume_manager_test.go:
```diff
@@ -0,0 +1,276 @@
+
+		manager, err := newTestVolumeManager(tmpDir, podManager, kubeClient)
+		if err != nil {
+			t.Fatalf("Failed to initialize volume manager: %v", err)
```
this should be `Errorf` and have a continue after it.

pmorie commented on pkg/securitycontext/provider_test.go:
```diff
@@ -181,32 +181,50 @@ func TestModifyHostConfigPodSecurityContext(t *testing.T) {
 	testCases := map[string]struct {
 		securityContext *api.PodSecurityContext
 		expected        *dockercontainer.HostConfig
+		extra           []int64
```
call this `extraSupplementalGroups`

pmorie commented:
Looks pretty good, just a couple nits in tests that I would like fixed.

pmorie commented:
@wongma7 squash commits please

Commit:
Remove pod mutation for PVs with supplemental GIDs

Commit:
Set extra supplemental groups for rkt

wongma7 commented:
@pmorie squashed feedbakc

childsb commented:
@k8s-bot test this issue #29473

pmorie commented:
@matchstick @saad-ali @eparis I'm marking this as a cherry pick candidate because I've spent a lot of time fighting bugs due to pod mutation; thoughts?

saad-ali commented:
That should be fine. This code has been baking for long enough and we haven't seen any issues.

pmorie commented:
Okay, cool.

fabioy commented:
@pmorie Cherrypick approved. Please create a cherrypick PR. Thanks.
